### PR TITLE
Implement websocket signalling for collaboration

### PR DESCRIPTION
**Pull request by nishantwrp**
Signed-off-by: Nishant Mittal <mittalnishant14@outlook.com>

**Description**
Future Todo: Currently there is no authentication in this endpoint. 

This PR fixes #

**Notes for Reviewers**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [X] Yes, I signed my commits.

### CODE REVIEW

File: server/handlers/pattern_collab_handler.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+	patterns                map[string]map[*Client]bool
+	broadcast               chan *SocketMessage
+	register                chan *Client
+	unregister              chan *Client
+	updateSubscribedPattern chan *PatternUpdateRequest
+}
+
+type SocketMessage struct {
+	Type   string      `json:"type"`
+	Topics []string    `json:"topics,omitempty"`
```
Comment (Revolyssup) [inline server/handlers/pattern_collab_handler.go:46]: what is the difference between `topics` and `topic` here?

File: server/handlers/pattern_collab_handler.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+var mainHub *Hub = nil
+
+func serveWs(hub *Hub, w http.ResponseWriter, r *http.Request) {
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		return
+	}
+
+	client := &Client{hub: hub, conn: conn, patternID: "", send: make(chan []byte), connected: true}
```
Comment (Revolyssup) [inline server/handlers/pattern_collab_handler.go:242]: why is patternid empty here?

File: server/handlers/pattern_collab_handler.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+		c.processMessage(message)
+	}
+}
+
+func (c *Client) writeMessages() {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+		c.hub.unregister <- c
```
Comment (Revolyssup) [inline server/handlers/pattern_collab_handler.go:115]: This same client was send to unregister channel when readMessages() returns. Also the connection is closed when the readMessages() returns. There will be a send() on closed channel. Do both these things at one place.

**Review by Revolyssup — CHANGES_REQUESTED** [review]

**Comment by Revolyssup** [comment]
@nishantwrp  All channels in hub are unbufferred. This might become a bottleneck so make these channels unbufferred unless some synchronization is required.

File: server/handlers/pattern_collab_handler.go (hunk 2 above)
Comment (nishantwrp) [reply server/handlers/pattern_collab_handler.go:242]: It is later sent by client when the type of message is "subscribe".

File: server/handlers/pattern_collab_handler.go (hunk 3 above)
Comment (nishantwrp) [reply server/handlers/pattern_collab_handler.go]: done!

**Comment by nishantwrp** [comment]
@Revolyssup got it. Made all the channels buffered. I don't think the communications need to be synchronous. I am unsure of the idea buffer length, so have made it `10` for hub channels and `100` for the client send channel. Let me know, if I should update that.

**Comment by leecalcote** [comment]
@nishantwrp when will this be ready for final review?

**Comment by nishantwrp** [comment]
@leecalcote it is ready for final review. // @Revolyssup

**Comment by leecalcote** [comment]
Syncing with `master`

**Comment by leecalcote** [comment]
Please update the REST API Endpoint spreadsheet and the REST API docs in Meshery Docs. (use make command)

**Pull request merged**